### ISSUE TITLE
What is SNT file and how to create a new file

### ISSUE THREAD

**Issue by axel578** [zhongkaifu/Seq2SeqSharp#52]
Hello !

I would like to know if it would be possible to create a new vocab snt file, I looked at this file in notepad but I'm note sure it's for vocabulary.

For my case I want to generate text which is some sort of xml file with  a limited number of token.
Every snt file available aren't suited to what I'm trying to do.

I would like to know if it's possible in any way to generate a new snt file based on a custom vocab

**Comment by GeorgeS2019** [comment]
@zhongkaifu 

> Different Text Generation Strategy: ArgMax, Beam Search, Top-P Sampling

Just curious, how is this language generation implemented similar or dissimilar to e.g. GPT-x ?

**Comment by zhongkaifu** [comment]
Hi @GeorgeS2019 

You could use Seq2SeqSharp to train GPT-x models only if you have training data set for it. They are all Transformer based model and data set is masked text.

**Comment by axel578** [comment]
Yes but what is .\spm\enuSpm.model for ? is it for the vocabulary, because on my scenario the vocabulary is a bunch of code different than the usual vocabulary

**Comment by zhongkaifu** [comment]
enuSpm.model is for SentencePiece to encode/decode for subword level tokens. Seq2SeqSharp can directly call APIs in SentencePiece for subword level encoding and decoding. SentencePiece has its own vocabulary in subword level, and it's different with your vocabulary. 

I don't think you need to care about it, because with parameters "-SrcSentencePieceModelPath"  and -TgtSentencePieceModelPath", Seq2SeqSharp can automatically encode word in your vocabulary to subword in model vocabulary, and decode subword back to word. With these two parameters, if you don't have vocabulary in subword level, you can set "SrcVocab" and "TgtVocab" to empty, and ask Seq2SeqSharp to generate vocabulary from training set. For inference, model itself already includes vocabulary.

**Comment by GeorgeS2019** [comment]
@zhongkaifu  

> You could use Seq2SeqSharp to train GPT-x models only if you have training data set for it

Is **importing trained weights from** e.g. GPT2 **.onnx** into Seq2SeqSharp model to avoid training is still part of a long term plan?

**Comment by GeorgeS2019** [comment]
@zhongkaifu 

> Different Text Generation Strategy: ArgMax, Beam Search, Top-P Sampling

I hope you are aware that: ArgMax, Beam Search, Top-P Sampling exist as OnnxRuntime's contributed operators

There is a need for these operators to stay [modular and reusable](https://github.com/zhongkaifu/Seq2SeqSharp/issues/39) for .NET 

Do you see the need for such? :-)

The ONNX industry community has already identified that and makes these Text Generation Strategy modular and reusable as contributed operators.

**Comment by zhongkaifu** [comment]
I know there is a need for modular and reusable, but it's not high priority and urgent for me right now. This is the reason why I say it's a long-term plan.
